org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/embedder/MavenImpl.java

```diff
@@ -1289,4 +1295,36 @@ public MavenExecutionResult execute(MavenExecutionRequest request) {
     }
   }
 
+  /**
+   * @param file a base file or directory, may be <code>null</code>
+   * @return the value for `maven.multiModuleProjectDirectory` as defined in Maven launcher
+   */
+  public static File computeMultiModuleProjectDirectory(File file) {
+    if(file == null) {
+      return null;
+    }
+    final File basedir = file.isDirectory() ? file : file.getParentFile();
+    File current = basedir;
+    while(current != null) {
+      if(new File(current, ".mvn").isDirectory()) {
+        return current;
+      }
+      current = current.getParentFile();
+    }
+    return basedir;
+  }
+
+  public MavenImpl cloneForBasedir(File basedir) {
```

Comment:
Why not passing an (optional) basedir to the createExecutionContext()? If we really need project specific maven impls then this hsould be handled in the IMavenFacade where you already added a `getMaven()` method

Reply:
Because we basically want several MavenImpl in the end, 1 per project; and not 1 MavenImpl with multiple contexts. Stateful is beautiful!
But yes, we can move it to MavenProjectFacade, I think the only reason why I left it here instead is because I was too lazy to add a setter for the new basedir.

Reply:
> Stateful is beautiful!

Yeah but the basedir is not really a state, and I think the idea is that we can share things across projects, an `IMaven` per projects would probably also blow up the memory consumption.

Reply:
> an IMaven per projects would probably also blow up the memory consumption.

Not for the moment as it's a shallow copy and things like ClassRealm (which are actually consuming RAM) are still shared.

Reply:
But that's the point here, if you just use IMaven to pass around a path, why not pass the path itself?
Especially as if an `IMaven` is bound to a project, where you could always get the basedir of the corresponding project instead...

Reply:
> But that's the point here, if you just use IMaven to pass around a path, why not pass the path itself?

Eventually, we'll pass more than the path. But the current issue is that it's much much more difficult to pass the path along the whole call chain as it requires to go through legacy APIs that do not allow propogation (don't accept extra args). I started trying that, had to create many more methods to IMaven, deprecate some more, it was basically not a good looking solution.

Reply:
I don't think we must pass much around, but in 
`ProjectRegistryManager.configureExecutionRequest(MavenExecutionRequest, IProjectRegistry, IFile, ResolverConfiguration)`

if there is a pom set the multimodule dir. In all other cases there won't be an executable project anyways?

Reply:
Again, I tried, and it's not as simple as you think. The ProjectRegistryManager is just the very most basic place where the execution request is created, and it's an easy one to tweak; but in many places, the chain is IMaven.createExecutionContext().getExecutionRequest(), and both those methods are API and one of them would need to be deprecated/replaced; which is something we are happy to avoid by having the "basedir" (or other project specific settings) included in the IMaven directly.

Reply:
> but in many places, the chain is IMaven.createExecutionContext().getExecutionRequest(), and both those methods are API and one of them would need to be deprecated/replaced; which is something we are happy to avoid by having the "basedir" (or other project specific settings) included in the IMaven directly.

But how is then than the basedir ever injected? if not through the project facade? Can you give one example of such call?

Reply:
In the end, we do need some specific IMaven (per project, or per group of project sharing the same .mvn, whatever), so what it the issue with using MavenProjectFacade.getMaven() to decide of the configuration of the IMaven? I really don't get what is the problem here, and what you're suggesting recommends to keep using singletons and tweaks 11 places where we can just tweak 3 and allow objects to carry more context and more customization, as it is functionally needed.

Reply:
> The issue is that you don't really have a specific IMaven you clone a static singleton just to pass a File object (for each call)

Still, it's a different Maven, that will allow futher configurations. How it's constructed, what's clones and what's not doesn't really matter in term of coupling from consumer end.

> constructing other objects later on this just feels wrong

I really don't get the same impression. On the other end, tweaking several methods to always receive the same arguments, or at least arguments that are coupled together is a sign that a dedicated instance can be worth it. I would suggest you refer to "Clean Code" and "Design Patterns" books that spend time on finding the smell of when singletons are ready to become instances. And this kind of "propagate 1 value in many places" is the main smell.

>  And given that this only works if I call method X because someone before has called Y in the same thread makes me thing that this is just waiting for bad things to happen (even though I can't proof it).

That threading stuff is indeed a hotspot for issues, but as far as I looked up, it doesn't cause any issue compared to currnet state.

> either there is already a proper set up context from the project manager anyways

Unfortunately, the request in project manager is not coupled to further invocations, so what happens in it has no impact on next calls to IMaven.